Forward port of Geant4 SIMP custom physics model to master branch

fabiocos opened the pull request:
#### PR description:

Port of #26094 (targeting 7_1_X) to master, based on the original proposal by @lowette .
The code has been adapted to the new 11_1_X branch trying to mimic as closely as possible the original proposal. The  ```G4GlauberGribovCrossSection``` class, no more present in Geant4, has been replaced by the available ```G4ComponentGGHadronNucleusXsc```. Include files have been moved to ```interface```, the ```G4``` names have been turned into ```CMS``` as done for many other models, and include guards have been updated according to the standard we adopt.

#### PR validation:

Code compiles and passes ```scram b code-checks```, but no dedicated test workflow is provided to probe the real code behaviour in the original proposal. @civanch should confirm the correctness of the adopted approach, and possible tests to be performed.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28763/13412

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - SimG4Core/CustomPhysics/src/G4QGSPSIMPBuilder.hh:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMP.hh:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMPInelasticXS.hh:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMPInelasticXS.cc:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMP.cc:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMPInelasticProcess.cc:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4SIMPInelasticProcess.hh:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4AntiSIMP.cc:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4QGSPSIMPBuilder.cc:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232
  - SimG4Core/CustomPhysics/src/G4AntiSIMP.hh:
    - Added: e0c01bd98919a09b482a00775c9385f62946022b
    - Deleted: 6c45bdec5d0af3eeecf1202f8c91faa816d6d232

cmsbuild commented:
A new Pull Request was created by  @fabiocos (Fabio Cossutti) for master.

It involves the following packages:

SimG4Core/CustomPhysics


@cmsbuild, @civanch, @mdhildreth can you please review it and eventually sign? Thanks.
@makortel this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fabiocos committed:
Add new files from lowette:simps_to_customphysics_7_1,
update files modified in lowette:simps_to_customphysics_7_1,
adapt SIMP model code to master, move to more standard file location

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28763/13413

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #28763 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 70a194911e6b245db21d22d5a8e6999b7aee7bc1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-308f52/4291/summary.html
CMSSW: CMSSW_11_1_X_2020-01-20-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-308f52/4291/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2697090
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2696743
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

civanch commented on SimG4Core/CustomPhysics/src/CMSQGSPSIMPBuilder.cc:
```diff
@@ -0,0 +1,57 @@
+  theStringModel->SetFragmentationModel(theStringDecay);
+
+  theCascade = new G4GeneratorPrecompoundInterface;
+  thePreEquilib = new G4PreCompoundModel(new G4ExcitationHandler);
```
@fabiocos , lines 22-23 should be removed.

civanch commented on SimG4Core/CustomPhysics/src/CMSQGSPSIMPBuilder.cc:
```diff
@@ -0,0 +1,57 @@
+CMSQGSPSIMPBuilder::~CMSQGSPSIMPBuilder() {
+  delete theStringDecay;
+  delete theStringModel;
+  delete thePreEquilib;
```
@fabiocos , lines 38-39 should be removed (delete thePreEquilib; delete theCascade)

civanch commented on SimG4Core/CustomPhysics/interface/CMSSIMPInelasticProcess.h:
```diff
@@ -0,0 +1,167 @@
+
+class G4ParticleDefinition;
+
+class CMSSIMPInelasticProcess : public G4HadronicProcess {
```
@fabiocos , most of virtual methods must be removed - they are implemented inside G4HadronicProcess. In different Geant4 releases they may be different.

civanch replied:
@fabiocos , The code was working in 10.0 where we run in single threading mode, in 10.6 many things were changed except interfaces, so the code is well compiling but there may be problems, which I cannot catch promptly.

civanch commented:
@fabiocos , would you agree if we merge this PR and after I will try to shape these classes to recent Geant4? I foresee in this 7_1 implementation some thread unsafe things not seen in an easy way. Suspect, that it would be ineffective if will put too many comments.

civanch commented:
+1
@fabiocos , I would think, that if we do forward port we should not change code a lot. So, let us integrate this but add modifications on top.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @silviodonato, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@civanch thanks for your comment, as said mine was a basic attempt to do the porting with the minimal needed code change. I might easily implement the first comments of yours, but if more is needed, may be it is better to follow your suggestion, let @silviodonato integrate this (it will not break anyway anything as far as I can see) and let you clean the code in a second step. Just to be sure, do we have some test workflow using CustomPhysics, just to have a small cross check?

silviodonato commented:
+1

smuzaffar commented:
@fabiocos , @civanch looks like this broke GEANT4 IBs 
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc820/CMSSW_11_1_GEANT4_X_2020-01-27-1100/SimG4Core/CustomPhysics

civanch commented:
@smuzaffar , thanks! 

I expect problems for GEANT4 IB but they happens on dramatically high level. Will try to make next PR asap. The reason of the problem: author of this code rewrite all Geant4 base class virtual methods for physics and x-sections with old variants of 6 years ago. Because he is using inheritance, he should rewrite instead only few methods, which really required his modifications.

fabiocos commented:
@civanch @smuzaffar as far as I can see, the issue to be solved in the code cleaning is the redefinition of the particle performed in https://github.com/cms-sw/cmssw/blob/master/SimG4Core/CustomPhysics/src/CMSSIMPInelasticProcess.cc#L145
If that could be moved inside the particle definition it would be straightforward to clean the CMSSIMPInelasticProcess class.
